### ISSUE TITLE
mqtt: /commands callback

### ISSUE THREAD

**Issue by sberryman** [nutechsoftware/AlarmDecoder-IoT#42]
Looking through the code I see you create a MQTT subscription but I can't find the callback.

https://github.com/nutechsoftware/AlarmDecoder-IoT/blob/master/components/ad2mqtt/ad2mqtt.cpp#L111-L117

Is this used at all? It would be great if we could use this to control the panel.

**Comment by f34rdotcom** [comment]
It is not but I am ready to wire it up and yes it needs to be done.  I had concerns for security so until I had some discussion with others on how best to provide security I left the callback out and just subscribe to the event. This update will be simple once this is brained and will be in the next release.

I was thinking a new mqtt command 
```mqtt secret some-randome-string```

Then the remote client must send in the secret as part of the JSON payload for the command to be accepted.

Thoughts?
 Best
 Sean M.

**Comment by sberryman** [comment]
I'm not sure how much security a token would provide to be honest. If someone is using MQTT (not SSL) they can easily intercept the token anyways. Also, the only security you are providing for ser2sock is IP address restrictions.

Maybe you disable `/commands` topic by default and when you enable it you also show a warning if the connection string is not behind a cert?

My personal setup is mosquitto running in the same vlan as all my other (private/internal) IoT devices with no routes to WAN. I also have ACL and username/password setup for the message server which is probably overkill.

My main driver for going the MQTT route is that home assistant's integration doesn't accurately detect the connection dropping (still not sure why the connection is getting severed but it is via wifi) and it gets stuck. The work around is restarting that integration every N minutes which is certainly helping.

Next up is home assistant auto-discovery. I know "auto" in this case isn't possible as you can't determine binary sensor device class for zone(s). This would require the user to set the zone device class which would be pretty easy to implement.

## Alarm panel:
https://www.home-assistant.io/integrations/alarm_control_panel.mqtt/
```yaml
# Example using numeric code with remote validation configuration.yaml
alarm_control_panel:
  - platform: mqtt
    name: "Alarm Panel With Numeric Keypad"
    state_topic: "alarmdecoder/panel"
    value_template: "{{ value_json.state }}"
    command_topic: "alarmdecoder/panel/set"
    code: REMOTE_CODE
    command_template: "{ action: '{{ action }}', code: '{{ code }}'}"
```

I like how home assistant's documentation makes it clear that MQTT is plain text 😁Did this used to be implemented as HA's documentation shows the topic of `alarmdecoder`

**Comment by sberryman-bird** [comment]
Thoughts?

**Comment by sberryman** [comment]
Thoughts?

**Comment by f34rdotcom** [comment]
After much but probably not enough thought I think a compromise is needed.

If the system is over https then I am not worried about the secret being exposed. I am also not worried about not requiring a key if connecting to a local MQTT broker. My concern is connecting to a public broker and anyone being able to access the system. I don't want to make it easy to create a problem. I suppose a warning when enabling MQTT about setting a key or by default setting some random key the user can later change would work. This would require you to remove the key or delete the key first. If the key is blank the command topic would allow the request with no provided key.

The template could look something like this.
`command_template: "{ secret: NA, action: '{{ action }}', code: '{{ code }}'}"`

If I am going to keep the MQTT client enabled I would be forced to generate a unique random secret for each device I ship. It would also then potentially be reported during the device booting over the console port.

```
logmode I
ad2source C 4:36
netmode E mode=d
webui enable Y
ser2sockd enable Y
ser2sockd acl 0.0.0.0/0
webui enable Y
webui acl 0.0.0.0/0
vpart 0 18
code 0 4112
mqtt url mqtt://mqtt.eclipseprojects.io/
#######
# Add the secret to the enable command to 
# require a secret attribute for the command
# topic.  Omit or leave blank if not provided will clear
# and disable checking for the secret attribute.
#mqtt enable Y <somesecret> 
mqtt enable Y

zone 209 Programming Mode
zone 095 Fire
zone 202 Alarm Cancel
```

**Comment by f34rdotcom** [comment]
After doing actual tests today I now see how far off I was on this. A public server for MQTT is not an option unless the data is also encrypted first. Then even if that happens a replay attack can be used so some sort of salt and rotating keys would need to be used. 

I will implement an enable / disable flag for the command sub and update dates to note the change. Expect to see this by this weekend time permitted this is a simple update.

**Comment by f34rdotcom** [comment]
@sberryman. Not tested needs sanity check.

Not sure how to deal with {UUID} Can I define custom variables on the YAML syntax such as ```uuid: "goosfraba...."``` and then refer to it in the templates by name ```{{uuid}}```.

I could add "exit mode" and more %if then% code. Would that be ```arming``` state? Arming but not fully armed yet. 


```
alarm_control_panel:
  - platform: mqtt
    name: "AlarmDecoder IoT partition 1"
    availability_topic: "ad2iot/{UUID}/status"
    availability_template: "{{ value_json.state }}"
    state_topic: "ad2iot/{UUID}/partitions/1"
    value_template: >
      {% if value_json.alarm_sounding == true || value_json.alarm_event_occurred == true %}triggered
      {% elif value_json.armed_stay == true %}
        {% if value_json.entry_delay_off == true %}armed_night
        {% elif value_json.entry_delay_off == false %}armed_home
        {% endif }
      {% elif value_json.armed_away == true %}
        {% if value_json.entry_delay_off == true %}armed_vacation
        {% elif value_json.entry_delay_off == false %}armed_away
        {% endif }
      {% else %}disarmed
      {% endif %}
    command_topic: "ad2iot/{UUID}/commands"
    code: REMOTE_CODE
    payload_arm_home: "ARM_STAY"
    payload_trigger: "PANIC_ALARM"
    command_template: "{ action: '{{ action }}', vpart: 0, code: '{{ code }}'}"
```

**Comment by f34rdotcom** [comment]
Initial work to add commands done and tested using MQTT Explorer. Tested basic ARM/DISARM using hand written json.
https://github.com/nutechsoftware/AlarmDecoder-IoT/commit/c673f32aec87a95b0b73081e58e363b4088ea271

**Comment by sberryman** [comment]
Well this is excellent news!

Panel yaml:
1. I'm assuming `ad2iot/{UUID}/status` is either {online | offline}?
2. UUID shouldn't be a problem if auto discovery is implemented.
3. I see you left `retain` with the default state of `False`. I suppose showing Unknown is more accurate than the previously know state for an alarm system though.
    * If retain is NOT set to true an update interval would be useful. Otherwise the state in home assistant will stay unknown until the zone or alarm state changes.

I think home assistant auto discovery should be a different component though. When enabling auto discovery you follow a very strict topic structure.

https://www.home-assistant.io/docs/mqtt/discovery/#discovery-topic
```
<discovery_prefix>/<component>/[<node_id>/]<object_id>/config

ex.
homeassistant/alarm_control_panel/{UUID}/config
```

This would allow creating the UUID on the ad2iot device and the user never needs to worry about it.

Autodiscovery component:

1. Enabled/Disabled
1. Topic prefix (should probably default to `homeassistant`)
1. Partitions, I only have 1 but I'm guessing your clients have systems with more than one...
1. Zone config (this is where it gets tricky)
    * Where is this defined? yaml or json file?
    * Do you configure them via MQTT?
    * Is there a GUI in the web app to do it?
    * Fortunately they don't change frequently

Edit: Really nice example of MQTT and home assistant auto discovery. https://github.com/tsightler/ring-mqtt

**Comment by f34rdotcom** [comment]
Still lots to process but here are a few answers. I need to dig into the UUID discovery option but if it can avoid the user having to type in the UUID of the device then that is a big win!

Panel yaml:
 #1. Correct. online/offline with a LWT so we will know if it is offline. Attached image example.

 #2. I think then a new command ```mqtt prefix``` that defaults to "" will work. So full path will be ```{prefix}/ad2iot/{uuid}/#``` This way users can set it to ```homeassistant``` and then the ad2iot topic will be created below that path. Max flexibility but not required still easy to setup.

 #3. Retain **should** be enabled on all publish calls? A client should be able to get everything about the alarm state as soon as they connect. If its not it needs fixing. I also set the default QOS level to 1.

Auto Discovery:
 #4 Zones are all under the zones tree see attached.
Other tree's exist but are not show here including.
 * ```cid```: Publish all contact ID reports from the panel if configured for LRR reporting to this topic.
 * ```switches```: This topic contains custom defined virtual switches configured by AD2IoT cli. Example would be to use this to monitor real-time state of an 5800RF sensor. Virtual switches can be created from all sorts of events or alarm panel states.

![image](https://user-images.githubusercontent.com/2042644/152717532-3d1fbd6c-51ed-4c0c-8ad2-9c902435eb9b.png)

**Comment by f34rdotcom** [comment]
New bug! Just tested setting my commands publish from my test mqtt client to enable retain and now whenever I reboot the AD2IoT it runs the same command again...

**Comment by sberryman** [comment]
1. great, I figured as you didn't include it in your example yaml. was just double checking
2. topic prefix should only be for home assistant auto configuration. you don't need that for your existing mqtt component
3. you don't want retain for any `commands`. you only retain for state updates. you might be doing that already to be honest
4. I saw you are publishing zone info. so this is where auto discovery comes in as being a separate component. you publish to `homeassistant/*/*/config` and in each of those you publish JSON (retain) with the configuration for the alarm panel and then each zone. in order to configure home assistant automatically you need to know what type of entity each zone is. Is it a smoke detector, a door, window, etc? So for each zone config you will tell home assistant to listen to `ad2iot/{UUID}/zones/{zoneid}` with a template to read the state. It's pretty simple but requires the ad2iot to know zone types.

**Comment by f34rdotcom** [comment]
#1  The yaml config I am testing uses availability_topic and template to detect if the AD2IoT is online/offline. I am working with limited knowledge and want to be sure my docs are correct for the HA community. I am 90% guessing :)
(https://github.com/nutechsoftware/AlarmDecoder-IoT/blame/sm-dev/README.md#L550)
    availability_topic: "ad2iot/{UUID}/status"
    availability_template: "{{ value_json.state }}"

#2 More info on this please. So in order to get auto detection where do I write this code? Is this something inside of HA I need to contribute some device info for everyone?

#3 Ok cool yes that is how its being done now and I now I am checking for retain flag on the commands topic to be sure I only get live data nothing recorded.

#4 This goes back to #2. I am reading up on HA MQTT discovery. It sounds like it would be best of my device sends these configs to the broker at startup. I presume for the ```alarm_control_panel``` I would include what I show as an example in README.md#L550?

**Comment by sberryman** [comment]
1. It seems like it is correct based on the reading I've done. I have NOT tried your build yet though
2. I would presume home assistant based auto detection would be a separate component enabled on the ad2iot. this would require zone configuration on the ad2iot to report the proper devices to homeassistant
3. excellent
4. correct, the only difference is the payload sent to configure the device is JSON. So in home assistant you enable MQTT and auto discovery. That is the only required configuration by a home assistant user within home assistant. 
5. When enabling home assistant discovery on the ad2iot you need to specify the home assistant mqtt topic (defaults to `homeassistant` and then you need to define a device class for each zone on your panel. then the ad2iot will send configuration messages to MQTT for each zone and the alarm panel itself. the panel and zones will read their state from your existing mqtt data on the `ad2iot` topic.

**Comment by f34rdotcom** [comment]
The config seems to work. Had to compress the value_template to a single line. Probably could be written more efficiently.

The prefix is optional since I can still specify the topic path(s) in the config json. Still a keeper feature. In this example I forced ```ad2iot``` below ```homeassistant``` by setting the prefix with the new command. I am working on adding either a specific ```homeassistant``` setting or a generic config publish for each partition configured and each zone configured on the AD2IoT. Not sure yet. Would like to keep it generic as much as possible.

```
{
  "unique_id": "AlarmDecoder_IoT_P1",
  "device_class": "alarm_control_panel",
  "name": "AlarmDecoder IoT P1",
  "availability_topic": "homeassistant/ad2iot/41443245-4d42-4544-4410-98f4ab2233c4/status",
  "availability_template": "{{ value_json.state }}",
  "state_topic": "homeassistant/ad2iot/41443245-4d42-4544-4410-98f4ab2233c4/partitions/1",
  "command_topic": "homeassistant/ad2iot/41443245-4d42-4544-4410-98f4ab2233c4/commands",
  "command_template": "{ \"vpart\": 0, \"action\": \"{{ action }}\", \"code\": \"{{ code }}\"}",
  "code": "REMOTE_CODE",
  "payload_arm_home": "ARM_STAY",
  "payload_trigger": "PANIC_ALARM",
  "value_template": "{% if value_json.alarm_sounding == true or value_json.alarm_event_occurred == true %}triggered{% elif value_json.armed_stay == true %}{% if value_json.entry_delay_off == true %}armed_night{% else %}armed_home{% endif %}{% elif value_json.armed_away == true %}{% if value_json.entry_delay_off == true %}armed_vacation{% elif value_json.entry_delay_off == false %}armed_away{% endif %}{% else %}disarmed{% endif %}"
}

```

![image](https://user-images.githubusercontent.com/2042644/153079787-f03f9cf9-82af-4c7e-8be4-0c0568888a73.png)

**Comment by f34rdotcom** [comment]
@sberryman Any suggestions for zone config for auto discovery. I am going to start digging into that next. I have inside of the IoT the zone # a zone description the user can set "Kitchen Motion" but they are all just ON/OFF so binary_sensory? A zone can also be a CO2 or Fire zone so how would that be dealt with? The AD2IoT does not know this info currently but it does know if a fire alarm is going off and that is usually showing up as Zone 96. I dont see any info on fire alarms inside of alarm_control_panel yet the data is available in the partition state on MQTT.

**Comment by sberryman** [comment]
@f34rdotcom this is so exciting that you are working on this! thank you!

I don't know the alarm panels enough to know all the different kinds of devices you can setup but all of mine are binary (doors, windows, motion/pir and smoke). 

So each zone would need to be enabled and device_class defined on the IoT. For my use case I have:
* door
* motion
* window
* smoke

Since zone 96 is fire (in my case) you can set that one up automatically with the user specifying the device class. I'm guessing you might have customers without fire (zone 96) so you should probably still require them to enable it?

Lets say zone 03 is a window:
```json
{
  "unique_id": "AlarmDecoder_IoT_P1_zone_003",
  "device_class": "window",
  "name": "AlarmDecoder IoT P1 Zone 003",
  "availability_topic": "homeassistant/ad2iot/41443245-4d42-4544-4410-98f4ab2233c4/status",
  "availability_template": "{{ value_json.state }}",
  "state_topic": "homeassistant/ad2iot/41443245-4d42-4544-4410-98f4ab2233c4/zones/003",
  "value_template": "{{ value_json.state }}",
  "payload_on": "OPEN",
  "payload_off": "CLOSE",
}
```

What is zone 252? I have that showing OPEN on my system.

**Comment by f34rdotcom** [comment]
Zone #'s as high as 252 does not sound like a 20P panel probably a 50PUL ? It is also possible that it is an AD2 firmware bug. I would like to see the raw messages from the panel where section 2 that has zone numbers shows 252. It is possible the 2 is a bad reported bit where it is zone 52 and some error. The ALPHA section hopefully will shed some light on the actual message.

Example where section 2 is 024 and alpha explains detail on it. If I can see the full message for 252 I should be able to figure it out.

```[00100001100000100A--],024,[f72600ff10240082280200000010f0],"CHECK 24 DOOR   LOCK            "```

Zones. On the AD2Iot firmware you can define all zones for the partition and then it will just post the config for those zones at startup.

```
AD2IOT # vpart 0
The vpart in slot 0 is 18 with a zone list of '4,5,6,7,8'
AD2IOT # zone 3
The alpha for zone 3 is 'some zone'
```

**Comment by f34rdotcom** [comment]
This seems to work well. I could add yet another sub command to ```mqtt``` to define the device type but it seems easy to just us a generic one and let users modify it in the configuration.yaml like this.

```
homeassistant:
  name: HOME
  unit_system: imperial
  customize:
    binary_sensor.ad2iot_zone_24:
      device_class: smoke
```

![image](https://user-images.githubusercontent.com/2042644/153532423-7b2657e4-e299-45ec-b4e1-26c0c6e49ee2.png)

![image](https://user-images.githubusercontent.com/2042644/153532545-3e4da0f7-e4f3-45ae-b09a-fdcdc9820548.png)

**Comment by sberryman** [comment]
I would agree with you if HomeAssistant let you change it in the UI. I try to avoid configuration file changes as much as possible now. If that is the route you go it will certainly work though. 

I also get the desire to leave it out of IoT. Hard decision on what is the best location for it. Personally I feel like it should live as close to possible of the actual device instead of pushing it down the line. But then again I'm going to be overriding the names of the zones anyways (via the UI, not a manual configuration file change)

Either way this is exciting progress!

What are your thoughts on MQTT and auto discovery so far? I'll be excited to upgrade and drop the existing integration. 

And I'll try and get you the raw output for the high zone this weekend. Haven't had a chance to do that yet. And it is a 20P.

**Comment by f34rdotcom** [comment]
I love the MQTT discovery! Having it just POP up and be visible is very cool. It helps me extract the parts I think are common and at the same time it documents how the data is extracted with the templates for others. Even if this is not a HA server it still provides good documentation on how to extract ON/OFF details from the partition document.

**Comment by sberryman** [comment]
The only thing I could think of is doing a binary sensor for `update_available`. This is what zigbee2mqtt does (creates an `binary_sensor.device_name_update_available`) for every single entity (light, switch, motion sensor, etc.)

This would at least allow the user in HA to create automations or display it in their UI if they want.

**Comment by f34rdotcom** [comment]
Ok FW update is done. I added a command FW_UPDATE but it is not wired to anything(yet). Still looking for how. Maybe a button with the action being the command topic with the FW_UPDATE content.

![image](https://user-images.githubusercontent.com/2042644/153685954-ac3f9546-c21b-4235-83e1-39001e884bc8.png)

**Comment by sberryman** [comment]
Looking great! I had no idea but you can actually auto configure a button and enable/disable that button based on whether a firmware update is available!

https://www.home-assistant.io/integrations/button.mqtt/

**Comment by f34rdotcom** [comment]
Fixed in release 1.0.9p4(#45)  I look forward to feedback and ideas on the next release. Only issue I left for now was not being able to set a device type for virtual mqtt switches. For now they are doors and can be changed via HA configuration. Next release I will plan on adding a new sub command to virtual switches with MQTT to define the device type. All Zones now have types but RF and virtual switches are outside of zones.

Best
 SM

**Comment by sberryman** [comment]
I finally updated the IoT device and it is working great! Thank you so much for implementing this!
